test/extended/util/test.go

```diff
@@ -147,7 +147,17 @@ func getNetworkSkips() []string {
 	return strings.Split(string(out), " ")
 }
 
-func newGinkgoTestRenamerFromGlobals(provider string, networkSkips []string) *ginkgoTestRenamer {
+func getOSSkips() []string {
+	out, err := e2e.KubectlCmd("get", "nodes", "--selector", "node-role.kubernetes.io/worker",
```

Comment:
No, you can’t run a command to get this info.

Detection would have to be done when the cloud provider info is filled out by reading from a key object.

It’s going to be better overall to just fix agnhost since this is a pretty key test and a future version of rhel could break.  As Casey says the error message is wrong, and that’s an indicator something about the environment is impacting the test.

Reply:
>Detection would have to be done when the cloud provider info is filled out by reading from a key object.

So read the AMI info instead? There is no other nice way to distinct OKD vs. OCP payload afaik.

Reply:
The info about the os should be on nodes.

I’m still inclined to say we need to fix the test.

Reply:
`os_id` label is set by MCO (as a [kubelet param](https://github.com/openshift/machine-config-operator/blob/master/templates/worker/01-worker-kubelet/_base/units/kubelet.yaml#L24)) so its seems to be a safe way of find out worker's OS.

Test definitely needs fixing, but its doesn't look its gonna get attention from SDN team :/

Reply:
> It’s going to be better overall to just fix agnhost

Casey's claim that it was agnhost's fault was wrong. Agnhost is correctly reporting that the test is not getting the expected result.

Vadim said on Slack that this is only failing on RHEL 7? In that case, I think it's a problem with RHEL 7 / legacy iptables mode then; and that explains why it was that it suddenly started working for us even though we hadn't changed anything. It didn't pass in 3.x but we thought it was because of openshift-sdn problems, so we had disabled it. But then when we tried re-enabling the test in 4.2, it just worked without us needing to change anything and we didn't know why. But it makes sense if the thing that changed was that OCP went from RHEL 7 to RHEL 8 underneath.

Reply:
This is reproducible on RHEL7 workers of OCP and Fedora CoreOS 31 on OKD.

I tried switching to nft tables on OKD ([replaced symlinks](https://github.com/openshift/release/commit/7e4dd6fcfa19bd537f56cb1c3139b84065158d9c)) but that didn't help much. It appears RHCOS (or just RHEL8) has some other secret fix / change

Reply:
I guess it doesn't really make sense that the difference would be iptables-legacy vs iptables-nft since we know that the test passes in upstream e2e with iptables-legacy.

And Fedora 31 has openvswitch 2.11 (which is what we're currently using in OCP) so it can't be OVS either...

Reply:
I'll try to summarize this issue:
The test was disabled because it wasn't passing.
Then _something_ happened and it started to pass again on RHCOS - however nobody seems to be able to pinpoint to what change exactly caused this.
The test was re-enabled, but continues to fail on RHEL7 and FCOS nodes.
Vadim and I have tried to solve this by making FCOS use iptables-nft instead of -legacy, but to no avail.
The SDN team is busy and won't have time to look into it this sprint.
This is now the main blocker for OKD 4 Beta 1.

I'd really like to move past this asap and disable the test temporarily until the proper solution is found, so we can get the OKD4 Beta out this week. 

@smarterclayton, the `getNetworkSkips` func above works similarly, and I agree with @vrutkovs that it seems to be a safe way to get the OS id from the node. If this is not the right way to exclude the test on FCOS, could you provide some more details on how we should do it instead?

Reply:
First, getNetworkSkips is also broken in the same way.  You cannot invoke Kubectl command at skip generation time.  All lookups like this need to be part of the default provider initialization, and use API calls to fetch info (if needed).

Second, this test is fundamental to how services are supposed to work (rejecting connections when there are no backends).  If it's broken, it's as if "pods don't start", and thus must be fixed.  A cluster that doesn't pass this isn't conformant to a Kube cluster.  It was temporarily skipped for OVN while they provided a fix, and should have already been removed.

Reply:
>If it's broken, it's as if "pods don't start", and thus must be fixed

Its rather "a different error message is thrown when a service with endpoints is used". It complicates the debugging, but its not as bad as "SDN doesn't route packages at all"

Reply:
Can we not skip this test exclusively on FCOS for now and make fixing this (as well as the getXSkips funcs) a blocker for OKD 4 GA? The change would only affect the OKD4 Beta, and with it being a Beta, I think we can get away with it not being entirely conformant in that regard.

Reply:
> Second, this test is fundamental to how services are supposed to work (rejecting connections when there are no backends). If it's broken, it's as if "pods don't start", and thus must be fixed.

The test name/description is somewhat misleading; what it's testing is "connections to a service with no backends are rejected immediately with 'connection refused' rather than timing out after several seconds with 'no route to host'". Everything still gets the expected results even when this test fails, it's just that when things are broken, clients will take longer to notice.

(This test was only added a year ago, at which point it was discovered that the "immediately reject" feature had been broken several months earlier in a kube-proxy refactoring, and no one had noticed.)

> A cluster that doesn't pass this isn't conformant to a Kube cluster.

This is not a `[Conformance]` test. (And before this test was added all of the other e2e tests had been passing for months while this feature was broken).

> It was temporarily skipped for OVN while they provided a fix, and should have already been removed.

It is still disabled for ovn-kubernetes (though it's expected to be fixed for 4.4).